File: Dynamic Programming/le152_maximumProductSubarray.py
"""
Find the contiguous subarray within an array (containing at least one number) which has the largest product.

For example, given the array [2,3,-2,4],
the contiguous subarray [2,3] has the largest product = 6.


"""


class Solution(object):
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums:
            return 0
        n = len(nums)
        localmin, localmax, globalmax = 1, 1, -0xffffffff
        for x in nums:
            localmax = max(localmax, 1) # keep localmax always 1 (prevent from 0 case)
            if x > 0:
                localmax, localmin = x*localmax, x*localmin
            else:
                localmax, localmin = x*localmin, x*localmax
            globalmax = max(globalmax, localmax)
        return globalmax

File: JavaPartI/le152_maximumProductSubarray.java
class Solution {
    public int maxProduct(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }
        int n = nums.length;
        int max = nums[0], min = nums[0], res = nums[0];
        for (int i = 1; i < n; i++) {
            int temp = min;
            min = Math.min(nums[i], Math.min(nums[i] * min, nums[i] * max));
            max = Math.max(nums[i], Math.max(nums[i] * temp, nums[i] * max));
            res = Math.max(res, max);
        }
        return res;
    }
}
